### PR TITLE
Statdb master db v3 848

### PR DESCRIPTION
**Pull request by aligeti**
Porting Statdb to master DB

### CODE REVIEW

File: pkg/statdb/statdb.go [hunk 1]
```diff
@@ -5,428 +5,145 @@ package statdb
...
-		}
-		invalidIds = append(invalidIds, id)
-	}
-
-	return &pb.FindInvalidNodesResponse{
-		InvalidIds: invalidIds,
-	}, nil
+// CreateResponse is a statdb create response message
+type CreateResponse struct {
+	Stats *pb.NodeStats
```
Comment (egonelbre) [inline pkg/statdb/statdb.go]: Return the actual results rather than some wrapper.

File: pkg/statdb/statdb_test.go [hunk 2]
```diff
@@ -5,461 +5,477 @@ package statdb_test
...
-	assert.NoError(t, err)
-
-	nodeID := teststorj.NodeIDFromString("testnodeid")
-
-	auditSuccessCount, totalAuditCount, auditRatio := getRatio(4, 10)
-	uptimeSuccessCount, totalUptimeCount, uptimeRatio := getRatio(8, 25)
-	err = createNode(ctx, db, nodeID, auditSuccessCount, totalAuditCount, auditRatio,
-		uptimeSuccessCount, totalUptimeCount, uptimeRatio)
-	assert.NoError(t, err)
+func TestIrreparable(t *testing.T) {
```
Comment (egonelbre) [inline pkg/statdb/statdb_test.go]: Test name error

File: pkg/statdb/statdb_test.go [hunk 3]
```diff
@@ -5,461 +5,477 @@ package statdb_test
...
-	if err != nil {
-		return &statdb.StatDB{}, &dbx.DB{}, err
-	}
-	db, err = dbx.Open("sqlite3", path)
-	if err != nil {
-		return &statdb.StatDB{}, &dbx.DB{}, err
-	}
-	return sdb, db, err
-}
+	t.Run("TestCreateDoesNotExist", func(t *testing.T) {
```
Comment (egonelbre) [inline pkg/statdb/statdb_test.go]: Let's not use subtests to keep test output cleaner.

File: satellite/satellitedb/dbx/satellitedb.dbx [hunk 4]
```diff
@@ -44,4 +44,29 @@ delete irreparabledb ( where irreparabledb.segmentpath = ? )
 read one (
   select irreparabledb 
   where  irreparabledb.segmentpath = ?
-)
\ No newline at end of file
+)
+
+// dbx.v1 golang statdb.dbx .
+model node (
```
Comment (mobyvb) [inline satellite/satellitedb/dbx/satellitedb.dbx]: I think you should change `node` to be called `statdb_node` or something. `node` is too generic for how we are using this model right now.

File: pkg/statdb/statdb.go [hunk 5]
```diff
@@ -5,428 +5,131 @@ package statdb
...
-		AND nodes.total_audit_count > 0
-		AND nodes.total_uptime_count > 0
-		AND (
-			nodes.audit_success_ratio < ?
-			OR nodes.uptime_ratio < ?
-		)`, args...)
-
-	return rows, Error.Wrap(err)
+// GetRequest is a statdb get request message
+type GetRequest struct {
```
Comment (mobyvb) [inline pkg/statdb/statdb.go]: `Get()` can just take `storj.NodeID` in as an argument. We do not need a struct

File: pkg/statdb/statdb.go [hunk 6]
```diff
@@ -5,428 +5,131 @@ package statdb
...
-		AuditSuccessRatio: dbNode.AuditSuccessRatio,
-		AuditCount:        dbNode.TotalAuditCount,
-		UptimeRatio:       dbNode.UptimeRatio,
-		UptimeCount:       dbNode.TotalUptimeCount,
-	}
-	return &pb.UpdateResponse{
-		Stats: nodeStats,
-	}, nil
+// GetResponse is a statdb get response message
+type GetResponse struct {
```
Comment (mobyvb) [inline pkg/statdb/statdb.go]: `Get()` can also just return a `*pb.NodeStats` instead of a struct

File: pkg/statdb/statdb.go [hunk 7]
```diff
@@ -5,428 +5,131 @@ package statdb
...
-		AuditCount:        dbNode.TotalAuditCount,
-		UptimeRatio:       dbNode.UptimeRatio,
-		UptimeCount:       dbNode.TotalUptimeCount,
-	}
-	return &pb.UpdateUptimeResponse{
-		Stats: nodeStats,
-	}, nil
+// FindInvalidNodesRequest is a statdb find invalid node request message
+type FindInvalidNodesRequest struct {
+	NodeIds  []storj.NodeID
```
Comment (mobyvb) [inline pkg/statdb/statdb.go]: same as `Get()`, take these as args instead of using the struct

File: pkg/statdb/statdb.go [hunk 8]
```diff
@@ -5,428 +5,131 @@ package statdb
...
-		auditSuccessCount,
-		totalAuditCount,
-	)
-
-	updateFields.AuditSuccessCount = dbx.Node_AuditSuccessCount(auditSuccessCount)
-	updateFields.TotalAuditCount = dbx.Node_TotalAuditCount(totalAuditCount)
-	updateFields.AuditSuccessRatio = dbx.Node_AuditSuccessRatio(auditRatio)
+// FindInvalidNodesResponse is a statdb find invalid node response message
+type FindInvalidNodesResponse struct {
+	InvalidIds []storj.NodeID
```
Comment (mobyvb) [inline pkg/statdb/statdb.go]: return `[]storj.NodeID` instead of `FindInvalidNodesResponse` struct

File: pkg/statdb/statdb.go [hunk 9]
```diff
@@ -5,428 +5,131 @@ package statdb
...
-		AuditSuccessRatio: dbNode.AuditSuccessRatio,
-		AuditCount:        dbNode.TotalAuditCount,
-		UptimeRatio:       dbNode.UptimeRatio,
-		UptimeCount:       dbNode.TotalUptimeCount,
-	}
-	return &pb.UpdateAuditSuccessResponse{
-		Stats: nodeStats,
-	}, nil
+// UpdateRequest is a statdb update response message
+type UpdateResponse struct {
```
Comment (mobyvb) [inline pkg/statdb/statdb.go]: `Update` should just return `*pb.NodeStats`

File: pkg/statdb/statdb.go [hunk 10]
```diff
@@ -5,428 +5,131 @@ package statdb
...
+	Stats *pb.NodeStats
 }
 
-// UpdateBatch for updating multiple farmers' stats in the db
-func (s *StatDB) UpdateBatch(ctx context.Context, updateBatchReq *pb.UpdateBatchRequest) (resp *pb.UpdateBatchResponse, err error) {
-	defer mon.Task()(&ctx)(&err)
+// UpdateUptimeRequest is a statdb uptime request message
+type UpdateUptimeRequest struct {
```
Comment (mobyvb) [inline pkg/statdb/statdb.go]: args instead of struct

File: pkg/statdb/statdb.go [hunk 11]
```diff
@@ -5,428 +5,131 @@ package statdb
...
+}
 
-	var nodeStatsList []*pb.NodeStats
-	var failedNodes []*pb.Node
-	for _, node := range updateBatchReq.NodeList {
-		updateReq := &pb.UpdateRequest{
-			Node: node,
-		}
+// UpdateUptimeResponse is a statdb uptime response message
+type UpdateUptimeResponse struct {
```
Comment (mobyvb) [inline pkg/statdb/statdb.go]: return `*pb.NodeStats` instead of struct

File: pkg/statdb/statdb.go [hunk 12]
```diff
@@ -5,428 +5,131 @@ package statdb
...
-		updateRes, err := s.Update(ctx, updateReq)
-		if err != nil {
-			s.log.Error(err.Error())
-			failedNodes = append(failedNodes, node)
-		} else {
-			nodeStatsList = append(nodeStatsList, updateRes.Stats)
-		}
-	}
+// UpdateAuditSuccessRequest is a statdb audit request message
+type UpdateAuditSuccessRequest struct {
```
Comment (mobyvb) [inline pkg/statdb/statdb.go]: args instead of struct

File: pkg/statdb/statdb.go [hunk 13]
```diff
@@ -5,428 +5,131 @@ package statdb
...
+	AuditSuccess bool
+}
 
-	updateBatchRes := &pb.UpdateBatchResponse{
-		FailedNodes: failedNodes,
-		StatsList:   nodeStatsList,
-	}
-	return updateBatchRes, nil
+// UpdateAuditSuccessResponse is a statdb audit response message
+type UpdateAuditSuccessResponse struct {
```
Comment (mobyvb) [inline pkg/statdb/statdb.go]: return `*pb.NodeStats` instead of struct

File: pkg/statdb/statdb.go [hunk 14]
```diff
@@ -5,428 +5,131 @@ package statdb
...
+	Stats *pb.NodeStats
 }
 
-// CreateEntryIfNotExists creates a statdb node entry and saves to statdb if it didn't already exist
-func (s *StatDB) CreateEntryIfNotExists(ctx context.Context, createIfReq *pb.CreateEntryIfNotExistsRequest) (resp *pb.CreateEntryIfNotExistsResponse, err error) {
-	defer mon.Task()(&ctx)(&err)
+// UpdateBatchRequest is a statdb update batch request message
+type UpdateBatchRequest struct {
```
Comment (mobyvb) [inline pkg/statdb/statdb.go]: take `NodeList` as an arg. Get rid of `UpdateBatchRequest`

File: pkg/statdb/statdb.go [hunk 15]
```diff
@@ -5,428 +5,131 @@ package statdb
...
-	}
-	if err != nil {
-		return nil, Error.Wrap(err)
-	}
-	createEntryIfNotExistsRes := &pb.CreateEntryIfNotExistsResponse{
-		Stats: getRes.Stats,
-	}
-	return createEntryIfNotExistsRes, nil
+// UpdateBatchResponse is a statdb update batch response message
+type UpdateBatchResponse struct {
```
Comment (mobyvb) [inline pkg/statdb/statdb.go]: return these without a struct

File: pkg/statdb/statdb.go [hunk 16]
```diff
@@ -5,428 +5,131 @@ package statdb
...
-		return 0, Error.Wrap(errs.New("success count less than 0"))
-	}
-	if totalCount < 0 {
-		return 0, Error.Wrap(errs.New("total count less than 0"))
-	}
-	if successCount > totalCount {
-		return 0, Error.Wrap(errs.New("success count greater than total count"))
-	}
+// CreateEntryIfNotExistsRequest is a statdb create entry request message
+type CreateEntryIfNotExistsRequest struct {
```
Comment (mobyvb) [inline pkg/statdb/statdb.go]: arg instead of struct

File: pkg/statdb/statdb.go [hunk 17]
```diff
@@ -5,428 +5,131 @@ package statdb
...
+	Node storj.NodeID
+}
 
-	ratio = float64(successCount) / float64(totalCount)
-	return ratio, nil
+// CreateEntryIfNotExistsResponse is a statdb create response message
+type CreateEntryIfNotExistsResponse struct {
```
Comment (mobyvb) [inline pkg/statdb/statdb.go]: return nodestats instead of struct

**Comment by mobyvb** [comment]
If you want, you can choose to keep all the structs, and I will fix them myself as part of my statdb refactor ticket. Especially if we just want to get this merged quickly so people can work on statdb again

File: cmd/overlay/cache.go [hunk 18]
```diff
@@ -49,10 +50,11 @@ func (c cacheConfig) open() (*overlay.Cache, error) {
 
 	// add logger
 	db = storelogger.New(zap.L(), db)
-	sdb, err = statdb.NewStatDB("postgres", dburl.String(), zap.L())
+
+	sdb, err := satellitedb.NewInMemory()
```
[the remark this answers is not in the document]
Comment (aligeti) [reply cmd/overlay/cache.go]: Done

File: cmd/captplanet/run.go [hunk 19]
```diff
@@ -129,7 +128,7 @@ func cmdRun(cmd *cobra.Command, args []string) (err error) {
 		errch <- runCfg.Satellite.Identity.Run(ctx,
 			grpcauth.NewAPIKeyInterceptor(),
 			runCfg.Satellite.Kademlia,
-			runCfg.Satellite.StatDB,
+			//runCfg.Satellite.StatDB,
```
Comment (mobyvb) [inline cmd/captplanet/run.go]: remove before merging

File: cmd/satellite/main.go [hunk 20]
```diff
@@ -123,7 +122,7 @@ func cmdRun(cmd *cobra.Command, args []string) (err error) {
 		grpcauth.NewAPIKeyInterceptor(),
 		runCfg.Kademlia,
 		runCfg.PointerDB,
-		runCfg.StatDB,
+		//runCfg.StatDB,
```
Comment (mobyvb) [inline cmd/satellite/main.go]: remove before merging

File: internal/testplanet/node.go [hunk 21]
```diff
@@ -155,12 +155,12 @@ func (node *Node) initOverlay(planet *Planet) error {
 	}
 	node.Kademlia = kad
 
-	dbPath := fmt.Sprintf("file:memdb%d?mode=memory&cache=shared", rand.Int63())
-	sdb, err := statdb.NewStatDB("sqlite3", dbPath, zap.NewNop())
+	sdb, err := satellitedb.NewInMemory()
```
[the remark this answers is not in the document]
Comment (aligeti) [reply internal/testplanet/node.go]: Done

File: cmd/satellite/main.go [hunk 22]
```diff
@@ -62,11 +61,11 @@ var (
...
-		Identity    provider.IdentityConfig
-		Kademlia    kademlia.Config
-		PointerDB   pointerdb.Config
-		Overlay     overlay.Config
-		StatDB      statdb.Config
+		Identity  provider.IdentityConfig
+		Kademlia  kademlia.Config
+		PointerDB pointerdb.Config
+		Overlay   overlay.Config
+		//StatDB      statdb.Config
```
[the remark this answers is not in the document]
Comment (aligeti) [reply cmd/satellite/main.go]: Done

File: cmd/captplanet/run.go [hunk 23]
```diff
@@ -39,15 +38,15 @@ const (
...
-	StatDB      statdb.Config
+	Identity  provider.IdentityConfig
+	Kademlia  kademlia.Config
+	PointerDB pointerdb.Config
+	Overlay   overlay.Config
+	Inspector inspector.Config
+	Checker   checker.Config
+	Repairer  repairer.Config
+	Audit     audit.Config
+	//StatDB      statdb.Config
```
[the remark this answers is not in the document]
Comment (aligeti) [reply cmd/captplanet/run.go]: Done

File: pkg/overlay/cache_test.go [hunk 24]
```diff
@@ -145,9 +146,13 @@ func TestCache_Store(t *testing.T) {
 	}
 	defer ctx.Check(planet.Shutdown)
 	planet.Start(ctx)
-	sdb := planet.Satellites[0].StatDB
 
-	testCache(ctx, t, teststore.New(), sdb)
+	sdb, err := satellitedb.NewInMemory()
```
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/overlay/cache_test.go]: done

File: pkg/overlay/config_test.go [hunk 25]
```diff
@@ -18,9 +17,9 @@ func TestRun(t *testing.T) {
 	var kadKey kademlia.CtxKey
 	ctx := context.WithValue(bctx, kadKey, kad)
 
-	sdb := &statdb.StatDB{}
-	var statKey statdb.CtxKey
-	ctx = context.WithValue(ctx, statKey, sdb)
+	// sdb, _ := satellitedb.NewInMemory()
+	// var statKey int //statdb.CtxKey
+	// ctx = context.WithValue(ctx, statKey, sdb.StatDB())
```
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/overlay/config_test.go]: Done

File: pkg/statdb/statdb.go (hunk 1 above)
Comment (aligeti) [reply pkg/statdb/statdb.go]: moby suggested he will work on refactoring all the structures

File: pkg/statdb/statdb_test.go (hunk 3 above)
Comment (aligeti) [reply pkg/statdb/statdb_test.go]: Tired to keep the changes to the minimal. Moby said he will work on it when he is refactoring statdb

File: cmd/satellite/main.go (hunk 20 above)
Comment (aligeti) [reply cmd/satellite/main.go]: done

File: cmd/captplanet/run.go (hunk 19 above)
Comment (aligeti) [reply cmd/captplanet/run.go]: done

File: pkg/pb/utils.go [hunk 26]
```diff
@@ -62,13 +62,9 @@ func CopyNode(src *Node) (dst *Node) {
 		}
 	}
 
-	node.AuditSuccess = src.AuditSuccess
-	node.IsUp = src.IsUp
 	node.LatencyList = src.LatencyList
```
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/pb/utils.go]: done

**Pull request merged**